c_code

```diff
@@ -564,3 +565,139 @@ for(i=0;i<8;i++)
+              return 1;
+         if(kingc-1>-1&&(board[kingr-2][kingc-1]*p==BKNIGHT))
+               return 1;
+     }
+     if(kingc+2<8)//KNIGHT
+       {
+           if(kingr+1<8 &&(board[kingr+1][kingc+2]*p==BKNIGHT))//not a friend
+               return 1;
+           if(kingr-1>-1&&(board[kingr-1][kingc+2]*p==BKNIGHT))
+               return 1;
+       }
+     if(kingc-2>-1)//KNIGHT
+       {
+          if(kingr+1<8 &&(board[kingr+1][kingc-2]*p==BKNIGHT))//not a friend
+               return 1;
+           if(kingr-1>-1&&(board[kingr-1][kingc-2]*p==BKNIGHT))
+               return 1;
+       }
+ i=kingr-1; //ROOK AND QUEEN
+while(i>-1)//bottom side
+       {
+       if(board[i][kingc]==SQUARE)
+           {
+           i--; continue;
+           }
+       if(board[i][kingc]*p==BROOK ||board[i][kingc]*p==BQUEEN)//first enemy
+           return 1;
+       break;
+       }
+   i=kingr+1;
+   while(i<8)//upper side
+       {
+       if(board[i][kingc]==SQUARE)
+           {
+           i++; continue;
+           }
+       if(board[i][kingc]*p==BROOK ||board[i][kingc]*p==BQUEEN)//first enemy
+           return 1;
+       break;
+       }
+   i=kingc-1;
+   while(i>-1)//left side
+   {
+   if(board[kingr][i]==SQUARE)
+       {
+        i--; continue;
+       }
+   if(board[kingr][i]*p==BROOK || board[kingr][i]*p==BQUEEN)
+       return 1;
+   break;
+   }
+   i=kingc+1;
+   while(i<8)//the right side
+   {
+   if(board[kingr][i]==SQUARE)
+       {
+           i++; continue;
+       }
+   if(board[kingr][i]*p==BROOK || board[kingr][i]*p==BQUEEN)
+      return 1;
+   break;
+   }
+ j=kingr+1; //BISHOP AND QUEEN
+  while(j<8 && kingc+j-kingr<8)//right-upper diaginal
+  {
+	 if(board[j][kingc+j-kingr]==SQUARE)
+	   {
+        j++; continue;
+       }
+	 if(board[j][kingc+j-kingr]*p==BBISHOP || board[j][kingc+j-kingr]*p==BQUEEN)
+		return 1;
+	 break;
+  }
+  j=kingr+1;
+  while(j<8 && kingc-j+kingr>-1)//left-upper diaginal
+  {
+	 if(board[j][kingc-j+kingr]==SQUARE)
+	   {
+        j++; continue;
+       }
+	 if(board[j][kingc-j+kingr]*p==BBISHOP || board[j][kingc-j+kingr]*p==BQUEEN)
+		return 1;
+	 break;
+  }
+  j=kingr-1;
+  while(j>-1 && kingc+j-kingr>-1)//left-bottom diaginal
+  {
+	 if(board[j][kingc+j-kingr]==SQUARE)
+	   {
+        j--; continue;
+       }
+	 if(board[j][kingc+j-kingr]*p==BBISHOP || board[j][kingc+j-kingr]*p==BQUEEN)
+		return 1;
+	 break;
+  }
+  j=kingr-1;
+  while(j>-1 && kingc-j+kingr<8)//right-bottom diaginal
+  {
+     if(board[j][kingc-j+kingr]==SQUARE)
+	   {
+        j--; continue;
+       }
+	 if(board[j][kingc-j+kingr]*p==BBISHOP || board[j][kingc-j+kingr]*p==BQUEEN)
+		return 1;
+	 break;
+  }
+  if(kingr+p<7 && kingr+p>0) //to validate the pawn option
+  {
+      if(kingc+1<8 && board[kingr+p][kingc+1]*p==BPAWN)
+        return 1;
+      if(kingc-1>-1 && board[kingr+p][kingc+1]*p==BPAWN)
+        return 1;
+  }
+  return 0;//all the attacked are protected or not there
+}
```

Comment:
you should add a check for the the enemy's king like this:
for(i=kingr-1;i<=kingr+1;i++)
    for(j=kingc-1;j<=kingc+1;j++)
  {
      if(i<0 || i>7 || j<0 || j>7) continue;
      if(board[i][j]*p==BKING) return 1;
  }